File: 4 занятие/3C_home.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    string st;
    cin >> st;

    stack<char> s;

    for (auto c : st) {
        if (c == '(' || c == '[' || c == '{') {
            s.push(c);
        } else if (s.size() == 0) {
            cout << "NO";
            return 0;
        } else {
            if (c == ')' && s.top() != '(' ||
                c == ']' && s.top() != '[' ||
                c == '}' && s.top() != '{') {
                    cout << "NO";
                    return 0;
            } else {
                s.pop();
            }
        }
    }
    if (s.size() != 0) {
        cout << "NO";
    } else {
        cout << "YES";
    }
}

File: 4 занятие/3C_home.py
import sys

s = input()
st = []

for c in s:
    if c in "([{":
        st.append(c)
    elif len(st) == 0:
        print("NO")
        sys.exit(0)
    else:
        if c == ")" and st[-1] != "(" or \
           c == "]" and st[-1] != "[" or \
           c == "}" and st[-1] != "{":
            print("NO")
            sys.exit(0)
        else:
            st.pop(-1)

if len(st) != 0:
    print("NO")
else:
    print("YES")
